src/snapred/backend/recipe/algorithm/ReductionAlgorithm.py

```diff
@@ -9,214 +9,129 @@
-        )
-
-        # TODO: Refactor so excute only needs to be called once
-        self.mantidSnapper.executeQueue()
 
-        dMin = {pgp.groupID: pgp.dResolution.minimum for pgp in reductionIngredients.pixelGroupingParameters}
-        dMax = {pgp.groupID: pgp.dResolution.maximum for pgp in reductionIngredients.pixelGroupingParameters}
-        dBin = {
-            pgp.groupID: pgp.dRelativeResolution / reductionIngredients.reductionState.instrumentConfig.NBins
-            for pgp in reductionIngredients.pixelGroupingParameters
-        }
-        groupIDs = [pgp.groupID for pgp in reductionIngredients.pixelGroupingParameters]
-        groupIDs.sort()
-        groupedData = data
-        for index, groupID in enumerate(groupIDs):
-            data = self.mantidSnapper.RebinRagged(
-                "Rebinning ragged bins...",
-                InputWorkspace=mtd[groupedData].getItem(index),  # or groupID
-                XMin=dMin[groupID],
-                XMax=dMax[groupID],
-                Delta=dBin[groupID],
-                OutputWorkspace="data_rebinned_ragged_" + str(focusGroups[index].name),  # or groupID
+        # prepare the data, vanadium with all same applications
+        for ws in [self.outputWorkspace, self.vanadiumWorkspace]:
+            self.mantidSnapper.MaskDetectors(
+                "Applying Pixel Mask...",
+                Workspace=ws,
+                MaskedWorkspace=self.maskWorkspace,
+            )
+            self.mantidSnapper.ApplyDiffCal(
+                "Applying Diffcal...",
+                InstrumentWorkspace=ws,
+                CalibrationWorkspace=self.calibrationTable,
+            )
+            self.mantidSnapper.ConvertUnits(
+                "Converting to Units of dSpacing...",
+                InputWorkspace=ws,
+                EMode="Elastic",
+                Target="dSpacing",
+                OutputWorkspace=ws,
+                ConvertFromPointData=True,
+            )
+            self.mantidSnapper.DiffractionFocussing(
+                "Applying Diffraction Focussing...",
+                InputWorkspace=ws,
+                GroupingWorkspace=self.groupingWorkspace,
+                OutputWorkspace=ws,
             )
+        # rebin vanadium to match data for division
+        self.mantidSnapper.RebinToWorkspace(
+            "Rebin vanadium to match",
+            WorkspaceToMatch=self.outputWorkspace,
+            WorkspaceToRebin=self.vanadiumWorkspace,
+            OutputWorkspace=self.vanadiumWorkspace,
+        )
+        # divide the data by the vanadium
+        # NOTE this must be done before ragged rebinning, since Divide cannot handle unequal axes
+        self.mantidSnapper.Divide(
+            "Dividing data by vanadium...",
+            LHSWorkspace=self.outputWorkspace,
+            RHSWorkspace=self.vanadiumWorkspace,
+            OutputWorkspace=self.outputWorkspace,
+        )
+        # ragged-rebin the data
+        self.mantidSnapper.RebinRagged(
```

Comment:
So this has been converted to operate on only one group at a time?

Reply:
For the time being.  I couldn't really tell what it was using before, but it seemed to be using the `GroupedWorkspace` of `GroupingWorkspace`s return by the `CustomGroupWorkspace` algorithm inside of mantid algorithms that require a `GroupingWorkspace`, and that didn't seem valid.

Assuming that was valid, then returning it to its former condition this should be as simple as passing it a `GroupedWorkspace` or `GroupingWorkspace`s inside of its own `GroupWorkspace` property.  If not, then when we refactor we can change this to take a `GroupedWorkspace` and iterate over all the workspaces in the group.

Reply:
Its probably better this way since we have moved away from GroupedWorkspaces